### PR TITLE
Add effectMap to Async

### PR DESCRIPTION
**Pull request by FireZenk**
#1484

- [x] Add function to Async.kt
- [x] Add dokka on top of the function (see other functions on the file)
- [x] Add examples below https://arrow-kt.io/docs/effects/async/#continueon

### CODE REVIEW

**Comment by FireZenk** [comment]
@nomisRev  @JorgeCastilloPrz

File: modules/docs/arrow-docs/docs/docs/effects/async/README.md [hunk 1]
```diff
@@ -87,6 +87,18 @@ Behind the scenes `continueOn()` starts a new coroutine and passes the rest of t
...
+
+Helper function that provides an easy way to construct a suspend effect
+
+```kotlin
+val userCreation = IO.async().run {
+  just(id)
+    .mapEffect { id -> createUserFromId(id) }
+}
+userCreation.request(it)
+```
```
Comment (nomisRev) [inline modules/docs/arrow-docs/docs/docs/effects/async/README.md]: This snippet is invalid because the type of `userCreation` is `Kind<ForIO, User>`.

**Comment by pakoito** [comment]
I suggested the snippet lol Any fix is good.

File: modules/docs/arrow-docs/docs/docs/effects/async/README.md (hunk 1 above)
Comment (FireZenk) [reply modules/docs/arrow-docs/docs/docs/effects/async/README.md]: Let's remove it then as suggested.

**Comment by pakoito** [comment]
New errors:

```
fun main(args: Array<String>) {
  //sampleStart
  suspend fun logAndIncrease(s: String): Int {
     println(s)
     return s.toInt() + 1
  }
  val result = ObservableK.async().mapEffect { s -> logAndIncrease(s) }
  //sampleEnd
  println(result)
}
```
error: unresolved reference: ObservableK
  val result = ObservableK.async().mapEffect { s -> logAndIncrease(s) }
               ^
error: cannot infer a type for this parameter. Please specify it explicitly.
  val result = ObservableK.async().mapEffect { s -> logAndIncrease(s) }
                                               ^
error: suspend function 'logAndIncrease' should be called only from a coroutine or another suspend function
  val result = ObservableK.async().mapEffect { s -> logAndIncrease(s) }
```

You have to add the relevant imports (check other examples), and hint `{ s: String ->`

**Comment by pakoito** [comment]
@FireZenk a refactor has landed, the class is now in `arrow-fx` so please move the code there :D

**Comment by pakoito** [comment]
Almost there! Add `* _imports_` like the prior snippets have!

**Comment by pakoito** [comment]
Bit closer:

```
import arrow.fx.rx2.extensions.observablek.applicative.*
```

Add the `async` instance here on top of the `applicative` one, that's what's missing.

**Comment by pakoito** [comment]
image

**Comment by pakoito** [comment]
The code is now correct, except there's a problem in codegen for suspend functions.

```
fun <A, B> Kind<ForIO, A>.effectMap(f: suspend A.() -> B): IO<B> = arrow.fx.IO.async().run {
  this@effectMap.effectMap<A, B>(f) as arrow.fx.IO<B>
}
```

**Comment by nomisRev** [comment]
@pakoito @FireZenk I'll take a look into the processor to see if we can fix this so we can include it in the release :)

**Comment by nomisRev** [comment]
I debugged this for a couple of hours and ran out of time. Having a hard time freeing up time to look into this. Created a [ticket](https://github.com/arrow-kt/arrow/issues/1554).

**Comment by raulraja** [comment]
I'll look into this

**Comment by raulraja** [comment]
Merging this into a local fix branch

**Pull request merged**